src/vdaf/mastic.rs

```diff
@@ -0,0 +1,651 @@
+}
+
+/// Mastic aggregation parameter.
+///
+/// This includes the VIDPF tree level under evaluation and a set of prefixes to evaluate at that level.
+#[derive(Clone, Debug)]
+pub struct MasticAggregationParam {
+    /// aggregation parameter inherited from [`Poplar1`]: contains the level (attribute length) and a vector of attributes (IdpfInputs)
+    poplar_param: Poplar1AggregationParam,
+    /// Flag indicating whether the VIDPF weight needs to be validated using SZK.
+    root_check_flag: bool,
+}
+
+impl Encode for MasticAggregationParam {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        self.poplar_param.encode(bytes)?;
+        let root_check = if self.root_check_flag { 1u8 } else { 0u8 };
+        root_check.encode(bytes)?;
+        Ok(())
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        Some(self.poplar_param.encoded_len()? + 1usize)
+    }
+}
+
+impl Decode for MasticAggregationParam {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        let poplar_param = Poplar1AggregationParam::decode(bytes)?;
+        let root_check = u8::decode(bytes)?;
+        let root_check_flag = root_check == 0;
+        Ok(Self {
+            poplar_param,
+            root_check_flag,
+        })
+    }
+}
+
+/// Mastic public share
+///
+/// Contains broadcast information shared between parties to support VIDPF correctness.
+pub type MasticPublicShare<V> = VidpfPublicShare<V>;
+
+impl<T, P, const SEED_SIZE: usize> ParameterizedDecode<Mastic<T, P, SEED_SIZE>>
+    for MasticPublicShare<VidpfWeight<T::Field>>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    fn decode_with_param(
+        mastic: &Mastic<T, P, SEED_SIZE>,
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        MasticPublicShare::<VidpfWeight<T::Field>>::decode_with_param(
+            &(mastic.bits, mastic.vidpf.weight_parameter),
+            bytes,
+        )
+    }
+}
+
+/// Mastic input share
+///
+/// Message sent by the [`Client`] to each Aggregator during the Sharding phase.
+#[derive(Clone, Debug)]
+pub struct MasticInputShare<F: FieldElement, const SEED_SIZE: usize> {
+    /// VIDPF key share.
+    vidpf_key: VidpfKey,
+
+    /// The proof share.
+    proofs_share: SzkProofShare<F, SEED_SIZE>,
+}
+
+impl<F: FieldElement, const SEED_SIZE: usize> Encode for MasticInputShare<F, SEED_SIZE> {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        self.vidpf_key.encode(bytes)?;
+        self.proofs_share.encode(bytes)?;
+        Ok(())
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        Some(self.vidpf_key.encoded_len()? + self.proofs_share.encoded_len()?)
+    }
+}
+
+impl<'a, T, P, const SEED_SIZE: usize> ParameterizedDecode<(&'a Mastic<T, P, SEED_SIZE>, usize)>
+    for MasticInputShare<T::Field, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    fn decode_with_param(
+        (mastic, role): &(&'a Mastic<T, P, SEED_SIZE>, usize),
```

Comment:
For consistency with the other `Vdaf` implementations
```suggestion
        (mastic, agg_id): &(&'a Mastic<T, P, SEED_SIZE>, usize),
```